spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -1143,66 +1147,64 @@ trait SnapshotManagement { self: DeltaLog =>
     }
   }
 
+  /** Get the snapshot at `version`. */
+  def getSnapshotAt(
+      version: Long,
+      lastCheckpointHint: Option[CheckpointInstance] = None): Snapshot = {
+    getSnapshotAt(version, lastCheckpointHint, lastCheckpointProvider = None)
+  }
+
   /**
-   * Get the snapshot at `version` using the given `lastCheckpointProvider` hint
+   * Get the snapshot at `version` using the given `lastCheckpointProvider` or `lastCheckpointHint`
    * as the listing hint.
    */
   private[delta] def getSnapshotAt(
       version: Long,
-      lastCheckpointProvider: CheckpointProvider): Snapshot = {
+      lastCheckpointHint: Option[CheckpointInstance],
+      lastCheckpointProvider: Option[CheckpointProvider]): Snapshot = {
+
     // See if the version currently cached on the cluster satisfies the requirement
-    val current = unsafeVolatileSnapshot
-    if (current.version == version) {
-      return current
+    val currentSnapshot = unsafeVolatileSnapshot
+    val upperBoundSnapshot = if (currentSnapshot.version >= version) {
+      // current snapshot is already newer than what we are looking for. so it could be used as
+      // upper bound.
+      currentSnapshot
+    } else {
+      val latestSnapshot = update()
```

Comment:
Do I understand correctly, that this `update` is needed in order to find the right commit store, in case the requested version is not yet backfilled?
 
(probably could use a doc comment, it's a subtle point)

Reply:
Yes.
Adding doc comment in a followup PR.